li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaConsumerImpl.java

```diff
@@ -319,6 +322,9 @@ public void unsubscribe() {
         }
       } catch (OffsetOutOfRangeException | NoOffsetForPartitionException oe) {
         handleInvalidOffsetException(oe);
+        if (_throwExceptionOnInvalidOffsets) {
+          ioe = oe;
```

Comment:
The exception can be thrown right here.  If we see OffsetOutOfRangeException here, rawRecords is guaranteed to be an empty map. So we don't need to wait until records are processed before throwing the exception.

Reply:
+1